osu.Game/Overlays/Music/PlaylistList.cs

```diff
@@ -98,10 +112,112 @@ public bool MatchingFilter
 
             public IEnumerable<IFilterable> FilterableChildren => Children;
 
-            public ItemSearchContainer()
+            public PlaylistListContainer()
             {
                 LayoutDuration = 200;
                 LayoutEasing = Easing.OutQuint;
+                privateList = new PlaylistLinkedList();
+            }
+
+            public Action<IList<PlaylistItem>> ReorderList;
+
+            private bool isItemBeingDragged;
+
+            private PlaylistItem itemBeingDragged;
+
+            private InputState stateDuringDrag;
+
+            private PlaylistLinkedList privateList;
+
+            public void OnChildDragStart(Drawable child, InputState state)
+            {
+                isItemBeingDragged = true;
+                itemBeingDragged = child as PlaylistItem;
+                stateDuringDrag = state;
+                return;
+            }
+
+            public void OnChildDrag(Drawable child, InputState state)
+            {
+                stateDuringDrag = state;
+                return;
+            }
+
+            public void OnChildDragEnd(Drawable child, InputState state)
+            {
+                isItemBeingDragged = false;
+                itemBeingDragged = null;
+                stateDuringDrag = null;
+                reorderList(child as PlaylistItem, state);
+                return;
+            }
+
+            protected override void UpdateAfterChildren()
+            {
+                base.UpdateAfterChildren();
+
+                if (isItemBeingDragged)
+                    itemBeingDragged.MoveToY(stateDuringDrag.Mouse.Position.Y);
+
+                return;
+            }
+
+            private void reorderList(PlaylistItem item, InputState state)
+            {
+                PlaylistItem pivot;
+                bool addToEnd = false;
+                int newPosition = (int) Math.Round(state.Mouse.Position.Y / item.Height);
+
+                if (newPosition >= privateList.Count)
+                {
+                    pivot = privateList[privateList.Count - 1];
+                    addToEnd = true;
+                }
+                else if (newPosition <= 0)
+                {
+                    pivot = privateList[0];
+                }
+                else
+                    pivot = privateList[newPosition];
+
+                privateList.Remove(item);
+
+                if (addToEnd)
+                    privateList.AddAfter(privateList.Find(pivot), item);

```

Comment:
This is not speeding things up at all. `Find` is a linear operation just like an insertion in a regular `List` would be. Only, that using linked lists you are hopping _pointers_, which cause frequent cache misses. Since cache misses are pretty much the most expensive thing that can happen to a processor, this is more like ~100x _slower_ than a regular `List` implementation which maintains a sorting.